hercules-ci-api/src/Hercules/Formats/CacheKeys/Keys.hs

```diff
@@ -0,0 +1,33 @@
+module Hercules.Formats.CacheKeys.Keys where
+
+import           Prelude
+
+import           Data.Aeson
+import           Data.Foldable
+import           Data.Text                      ( Text )
+
+-- | Credentials and public keys for a cache.
+data Keys = Keys
+ { signingKeys :: [Text]
+ , pullToken :: Maybe Text
+ , publicKeys :: [Text]
+ }
+
+instance ToJSON Keys where
+  toJSON a =
+    object
+      $ ["signingKeys" .= signingKeys a]
+      <> foldMap (pure . ("pullToken" .=)) (toList $ pullToken a)
```

Comment:
If we go with netrc authentication terminology, it should be `password`, but it isn't a password because it's not supposed to be remembered etc.
This naming is indeed a remnant of the Cachix-like terminology that was used, but I've started making changes to generalize it, because this data structure can be used by other caches in the future. This is a benefit of not using a Cachix-specific export format.
So perhaps it *should* be called password, despite that it doesn't look pretty for the specific case of Cachix. After all that's the term that generalizes when we add support for usernames.

(Historical note: suppose we did go with Cachix terminology, it's currently an `authtoken` (as in `cachix authtoken`). However, in the future it will not have to be *that* token. Instead it can be any token from Cachix, whatever it may be called, as long as the token is authorized for pulling, because that's how the agent needs to use it. How that works exactly doesn't matter for the agent. It just needs to pull. This is why the `pullToken` name made sense to me.)

Reply:
`password` is an FTP term, while netrc configures basic auth when performing HTTP.

Note that in the future binary cache could be custom, so it's better to stick with Nix/HTTP terminology.

How the token is called on Cachix side is still debatable so I'd like to eliminate that debate now and use the general term that is widely applicable.

Reply:
Hm, right, so curl allows us to abuse the field by not providing the username and colon? ([rfc, page 4](https://tools.ietf.org/html/rfc7617)) Or is username an empty string?

Not sure what's going on but providing a netrc `login` field (the username) is a valid thing to do. The Basic Auth RFC talks about "user-id" most of the time, but also "username" once when talking about the user agent. "login" is only used for the act of logging in. So, plenty of choice :(. I'd go with username, because I believe that to be the most common term generally. Even the [FTP](https://tools.ietf.org/html/rfc959) uses that, so I don't know where netrc gets "login" from. Perhaps an interpretation of "login information" (== "credentials").

Reply:
https://github.com/cachix/cachix/blob/master/cachix/src/Cachix/Client/NetRc.hs#L54

Reply:
This file format doesn't need to be Cachix-specific. `netrc` is merely a means to an end, so I'd like to hide its wrong naming from the user. So preferably the instructions will be like: put the Cachix authToken value in the `basicAuthPassword` field. That makes for a small and accurate public interface.

Suppose we do pick Cachix-specific terms, we'll end up in a situation where users with non-Cachix caches have to put their password in the `authToken` field (which seems wrong), or we need to duplicate the field (`authToken` + `basicAuthPassword`) just to make it look less wrong.

Re `basicAuthLogin`: scope is Cachix only for now, so we can wait with `basicAuthUserId` or `basicAuthUsername` field.

Reply:
That's reasonable :)

Seems then we have two pathways:

a) go with something like nominal typing of JSON with `type: cachix` and `authToken: ...`

b) abstract over the binary cache interface and use low-level terms like netrc

Thinking further it seems like we'll need to have `url` when going non-cachix and cachix will most likely need new options that will be cachix specific.

Given we went with `kind` it seems best to be consistent with (a)

Reply:
Sgtm.

If we're going to add a type/tag/kind/you-name-it, the outer `kind` field becomes redundant. Kind-tagging them individually seems like a good idea because it makes each cache kind's code independent and independently usable.
Example:
```
{ "foo":
    { "kind": "cachixCache",
      "authToken": ...,
      "signingKeys": ...,
      "publicKeys": ...
    },
  "my-custom-cache":
    { kind: "httpCache",
      ...
    }
}
```

Notes
 - the cachix cache name is implied by the Map-like field name.
 - some non-confidential fields are included. Atomic config operations are good for security.